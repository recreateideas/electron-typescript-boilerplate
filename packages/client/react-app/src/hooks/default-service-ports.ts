import { IServicePorts } from '../typings';

export const defaultServicePorts = {
    'local-service': 2999,
} as IServicePorts;
