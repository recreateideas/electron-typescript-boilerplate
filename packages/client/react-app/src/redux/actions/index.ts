import * as common from './common/actions';
import * as router from './router/actions';
import * as theme from './theme/actions';

export { common, router, theme };
