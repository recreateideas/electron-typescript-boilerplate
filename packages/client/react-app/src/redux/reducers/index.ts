import common from './common';
import theme from './theme';

const reducers = {
    common,
    theme,
};

export default reducers;
