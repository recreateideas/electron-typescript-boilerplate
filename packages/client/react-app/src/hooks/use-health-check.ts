import { useState, useEffect } from 'react';
import { selectors, useSelector } from '../redux';
import { servicesHealthcheck } from '../utils';
import { IServicePorts } from '../typings';

interface IUseHealthChecks {
    servicePorts: IServicePorts | undefined;
}

const useHealthCheck = ({ servicePorts }: IUseHealthChecks): boolean | undefined => {
    const [isHealthy, setIsHealthy] = useState<boolean | undefined>();
    const { common: commonSelectors } = selectors;
    const isElectron = useSelector(commonSelectors.isElectron);
    useEffect(() => {
        if (servicePorts) {
            const healthCheck = async () => {
                const result = await servicesHealthcheck({
                    servicePorts,
                    isElectron,
                });
                setIsHealthy(result);
            };
            healthCheck();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [servicePorts, isElectron]);
    return isHealthy;
};

export default useHealthCheck;
