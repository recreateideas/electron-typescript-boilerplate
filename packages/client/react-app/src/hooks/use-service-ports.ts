import { useEffect } from 'react';
import { actions, useDispatch, selectors, useSelector } from '../redux';
import { IServicePorts } from '../typings';
import { defaultServicePorts } from './default-service-ports';

const electron = { ...(window.require ? window.require('electron') : {}) };

const { ipcRenderer } = electron;

const useServicePorts = (): undefined | IServicePorts => {
    const dispatch = useDispatch();
    const {
        common: { getServicePorts, setServicePorts, setIsElectron },
    } = actions;
    const { common: commonSelectors } = selectors;
    const servicePorts: undefined | IServicePorts = useSelector(commonSelectors.servicePorts);
    const isElectron = !!ipcRenderer;
    useEffect(() => {
        dispatch(setIsElectron(isElectron));
        /* istanbul ignore if */
        if (isElectron) {
            dispatch(getServicePorts());
        } else {
            dispatch(setServicePorts(defaultServicePorts));
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    return servicePorts;
};

export default useServicePorts;
