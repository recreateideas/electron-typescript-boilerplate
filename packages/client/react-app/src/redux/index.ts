import { useDispatch } from 'react-redux';
import * as selectors from './selectors';
import * as actions from './actions';
import { useSelector } from './custom-hooks';

export * from './store';

export { useDispatch, actions, useSelector, selectors };
