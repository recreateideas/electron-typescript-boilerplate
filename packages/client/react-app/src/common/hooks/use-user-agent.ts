import { useState, useEffect, useMemo } from 'react';
import { UserAgent } from 'src/redux';
//@ts-ignore
import useMobileDetect from 'use-mobile-detect-hook';

export const useUserAgent = (): UserAgent | undefined => {
    const { isMobile, isDesktop, isAndroid, isIos } = useMobileDetect();
    const [ua, setUA] = useState<UserAgent | undefined>();
    useEffect(() => {
        const getUA = () => {
            const breakpoint = 620;
            const userAgent: UserAgent = {
                isMobile: isMobile() || window.innerWidth <= breakpoint,
                isDesktop: isDesktop(),
                isAndroid: isAndroid(),
                isIos: isIos(),
            };
            setUA(userAgent);
        };
        window.addEventListener('resize', getUA);
        getUA();
        return () => {
            window.removeEventListener('resize', getUA);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    return useMemo(() => ua, [ua?.isMobile]);
};
