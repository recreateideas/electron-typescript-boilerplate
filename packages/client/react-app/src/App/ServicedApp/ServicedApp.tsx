import { ReactElement } from 'react';
import { useHealthCheck, useServicePorts } from '../../hooks';
import { Loader } from '../ui-core';
import { App } from '../App';

export const ServicedApp = (): ReactElement => {
    const servicePorts = useServicePorts();
    const isHealthy = useHealthCheck({ servicePorts });
    const isChecking = isHealthy === undefined;
    return (
        <div id="app">
            {isHealthy ? (
                <App />
            ) : !servicePorts ? (
                <Loader
                    {...{
                        className: 'setting-up',
                        color: 'green',
                        message: 'setting up services...',
                    }}
                />
            ) : isChecking ? (
                <Loader
                    {...{
                        className: 'checking',
                        color: '#6b5b95',
                        message: 'checking services...',
                    }}
                />
            ) : (
                <Loader
                    {...{
                        className: 'not-responding',
                        color: 'red',
                        freeze: true,
                        message: 'Services are not responding. Cmd + R to retry.',
                    }}
                />
            )}
        </div>
    );
};
