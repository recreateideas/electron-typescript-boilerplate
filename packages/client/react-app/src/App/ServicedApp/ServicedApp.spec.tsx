import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';
import { store } from '../../redux';
jest.mock('../../hooks');
import * as hooks from '../../hooks';
import { ServicedApp } from './ServicedApp';

describe('App', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <Provider store={store}>
                <ServicedApp />
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });
    it('should match snapshot', () => {
        const subject = mount(
            <Provider store={store}>
                <ServicedApp />
            </Provider>
        );
        expect(EnzymeToJson(subject)).toMatchSnapshot();
    });

    it('should render .setting-up loader when service-ports are undefined', () => {
        jest.spyOn(hooks, 'useServicePorts').mockImplementation(() => undefined);
        const wrapper = mount(
            <Provider store={store}>
                <ServicedApp />
            </Provider>
        );
        const Loader = wrapper.exists('.setting-up');
        expect(Loader).toBeTruthy();
    });
    it('should render .checking loader when isHealthy is undefined', () => {
        jest.spyOn(hooks, 'useServicePorts').mockImplementation(() => ({ 'some-service': 1111 }));
        jest.spyOn(hooks, 'useHealthCheck').mockImplementation(() => undefined);
        const wrapper = mount(
            <Provider store={store}>
                <ServicedApp />
            </Provider>
        );
        const Loader = wrapper.exists('.checking');
        expect(Loader).toBeTruthy();
    });
    it('should render .not-responding loader when service-ports are loaded and isHealthy is false', () => {
        jest.spyOn(hooks, 'useServicePorts').mockImplementation(() => ({ 'some-service': 1111 }));
        jest.spyOn(hooks, 'useHealthCheck').mockImplementation(() => false);
        const wrapper = mount(
            <Provider store={store}>
                <ServicedApp />
            </Provider>
        );
        const Loader = wrapper.exists('.not-responding');
        expect(Loader).toBeTruthy();
    });
});
