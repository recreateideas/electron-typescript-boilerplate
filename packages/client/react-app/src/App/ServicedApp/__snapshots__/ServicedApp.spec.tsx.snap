// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServicedApp/> should match snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ServicedApp>
    <div
      id="app"
    >
      <Loader
        className="setting-up"
        color="green"
        message="setting up services..."
      >
        <styled.div
          className="loader setting-up"
        >
          <div
            className="sc-bdnxRM ivVdKc loader setting-up"
          >
            <Spinner
              color="green"
              radius={40}
              stroke={5}
              visible={true}
            >
              <div
                className="react-spinner-material"
                style={
                  Object {
                    "borderColor": "green",
                    "borderWidth": 5,
                    "height": 40,
                    "width": 40,
                  }
                }
              />
            </Spinner>
            <styled.div>
              <div
                className="sc-gtsrHT iiOSuX"
              >
                setting up services...
              </div>
            </styled.div>
          </div>
        </styled.div>
      </Loader>
    </div>
  </ServicedApp>
</Provider>
`;
